src/react/components/common/tagInput/tagInput.tsx

```diff
@@ -0,0 +1,439 @@
+        const { editingTag } = this.state;
+        this.setState({
+            editingTag: editingTag && editingTag.name === tag.name ? null : tag,
+            clickedColor,
+            showColorPicker: !this.state.showColorPicker && clickedColor,
+        });
+    }
+
+    private handleClick = (tag: ITag, props: ITagClickProps) => {
+        if (props.ctrlKey && this.props.onCtrlTagClick) {
+            this.props.onCtrlTagClick(tag);
+            this.setState({ clickedColor: props.clickedColor });
+        } else if (props.altKey) {
+            this.onAltClick(tag, props.clickedColor);
+        } else {
+            const { editingTag, selectedTag } = this.state;
+            const inEditMode = editingTag && tag.name === editingTag.name;
+            const alreadySelected = selectedTag && selectedTag.name === tag.name;
+
+            this.setState({
+                editingTag: inEditMode ? null : editingTag,
+                selectedTag: (alreadySelected && !inEditMode) ? null : tag,
+                clickedColor: props.clickedColor,
+                showColorPicker: false,
+            });
+
+            // Only fire click event if a region is selected
+            if (this.props.selectedRegions &&
+                this.props.selectedRegions.length > 0 &&
+                this.props.onTagClick &&
+                !inEditMode) {
+                this.props.onTagClick(tag);
+            }
+        }
+    }
+
+    private deleteTag = (tag: ITag) => {
+        if (!tag) {
+            return;
+        }
+        const index = this.state.tags.indexOf(tag);
+        const tags = this.state.tags.filter((t) => t.name !== tag.name);
+        this.setState({
+            tags,
+            selectedTag: this.getNewSelectedTag(tags, index),
+        }, () => this.props.onChange(tags));
+        if (this.props.lockedTags.find((l) => l === tag.name)) {
+            this.props.onLockedTagsChange(
+                this.props.lockedTags.filter((lockedTag) => lockedTag !== tag.name),
+            );
+        }
+    }
+
+    private getNewSelectedTag = (tags: ITag[], previouIndex: number): ITag => {
+        return (tags.length) ? tags[Math.min(tags.length - 1, previouIndex)] : null;
+    }
+
+    private onSearchKeyDown = (event: KeyboardEvent): void => {
+        if (event.key === "Escape") {
+            this.setState({
+                searchTags: false,
+            });
+        }
+    }
+
+    private onAddTagKeyDown = (event) => {
+        if (event.key === "Enter") {
+            // validate and add
+            const newTag: ITag = {
+                name: event.target.value,
+                color: this.getNextColor(),
+            };
+            if (newTag.name.length && !this.state.tags.find((t) => t.name === newTag.name)) {
+                this.addTag(newTag);
+                event.target.value = "";
+            } else {
+                // toast that tells them to pick another name
```

Comment:
Is it okay that this isn't implemented yet?

Reply:
I'll see if I can add it real quick